### ISSUE TITLE
Converter should use Jackson kotlin module if available when mapping JSON to kotlin data classes

### ISSUE THREAD

**Issue by saibotma** [jOOQ/jOOQ#11165]
### Actual behavior
When using SQL/JSON and fetching into a Kotlin data class then the mapping fails with the following exception:
`com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of Table (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
`

According to an [answer on SO](https://stackoverflow.com/a/53192674/6172447) this is because Jackson requires https://github.com/FasterXML/jackson-module-kotlin in order to map to Kotlin data classes.

A workaround to this issue is to register a custom `ConverterProvider` that uses Jackson with the Kotlin module.
I have taken this snippet from https://www.jooq.org/doc/3.14/manual/sql-execution/fetching/converter-provider/ and modified it to use the Kotlin Jackson module (or in the following example allows me to pass it to the constructor).
```
class JacksonKotlinConverterProvider(private val objectMapper: ObjectMapper) : ConverterProvider {
    private val delegate: ConverterProvider = DefaultConverterProvider()

    override fun <T : Any?, U : Any?> provide(tType: Class<T>?, uType: Class<U>?): Converter<T, U>? {
        if (tType == JSON::class) {
            return Converter.ofNullable(tType as Class<T>, uType,
                { t: T ->
                    try {
                        return@ofNullable objectMapper.readValue((t as JSON).data(), uType)
                    } catch (e: Exception) {
                        throw DataTypeException("JSON mapping error", e)
                    }
                },
                { u: U ->
                    try {
                        val w = StringWriter()
                        val g = JsonFactory().createGenerator(w)
                        objectMapper.writeValue(g, u)
                        return@ofNullable JSON.valueOf(w.toString()) as T
                    } catch (e: Exception) {
                        throw DataTypeException("JSON mapping error", e)
                    }
                }
            )
        } else
            return delegate.provide(tType, uType)
    }
}
```

One downside of this solution (or maybe not just of this solution) is that all the custom `Converter`s I created previously are not being used in this converter provider anymore.
For example, I have written a `Converter` in order to have `Instant` instead of `OffsetDateTime` and I had to register/write an extra Jackson module that parses date time strings in the JSON coming from the database to `OffsetDateTime` and then get an `Instant` from it.

### Expected behavior

I would expect jOOQ to use the Kotlin Jackson module by default.
However, I'm not sure if its jOOQ responsibility to solve my issue with the custom `Converters`, would be great though if jOOQ would just magically handle it correctly.

### Steps to reproduce the problem
- Use jOOQ in Kotlin
- Create a simple data class
- Use SQL/JSON to get JSON and try to map it to the data class by using `fetchInto`

### Versions

- jOOQ: 3.14.4
- Java: 11
- Database (include vendor): PostgreSQL 12
- OS: macOS
- JDBC Driver (include name if inofficial driver): 42.2.18

**Comment by lukaseder** [comment]
Thanks a lot for your report. That sounds like two separate issues:

1. jOOQ auto-loading that Kotlin Jackson module and using it correctly
2. `ConverterProvider` not interfering with other converters

Can you show an example of how the latter can be reproduced? Perhaps, create a new issue for the latter...

**Comment by saibotma** [comment]
I'm sorry, but I am very busy at the moment. 
I have this ticket on my to-do list and will show you more examples asap.

**Comment by lukaseder** [comment]
Do you know if this is still a problem, @saibotma?

**Comment by saibotma** [comment]
As far as I know @JulianLenz_ had the same issue when setting up his new project some weeks ago. Do you know which jOOQ version you used @JulianLenz_?

I have to admit that I did not try again if the issue is gone.

**Comment by JulianBissekkou** [comment]
We are using `3.14.9` 👍🏽

**Comment by strindberg** [comment]
Problem 1 is most definitely still a problem. Without proper support for Kotlin Jackson, the whole object mapping support is close to unusable in Kotlin.

**Comment by lukaseder** [comment]
Alright thanks for the feedback. I'll review part 1 of this issue, ignoring the `ConverterProvider` interference problems mentioned in the workaround description. If that still pops up, please open a new issue.

**Comment by strindberg** [comment]
If it's of any use, I've made a minimal, runnable example here: https://github.com/strindberg/jooq-issue-11165

With these table definitions:
```sql
CREATE TABLE parent (
    id     bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY,
    name   text   NOT NULL,
    PRIMARY KEY (id)
);

CREATE TABLE child (
    name                 text         NOT NULL,
    parent_id         bigint,
    FOREIGN KEY (parent_id) REFERENCES parent (id)
);
```

And this Kotlin code:
```kotlin
dsl.select(
            listOf(
                PARENT.ID,
                PARENT.NAME,
                field(
                    select(jsonObject(CHILD.NAME))
                        .from(CHILD).where(
                            CHILD.PARENT_ID.eq(PARENT.ID)
                        )
                ).`as`("child"),
            )
        )
            .from(PARENT).where(PARENT.ID.eq(id)).fetchOne()?.into(Parent::class.java)
```

The following exception is the result:
```
com.fasterxml.jackson.databind.exc.MismatchedInputException: 
    Cannot construct instance of `se.jh.jooqissue.repo.Child` 
    (although at least one Creator exists): cannot deserialize from Object value (no delegate- or property-based Creator)
```

**Comment by lukaseder** [comment]
Thanks, @strindberg. I can't run the example:

```
lukas@LAPTOP-VH130UJV MINGW64 ~/jOOQ/mcve/11165/jooq-issue-11165
$ ./gradlew build
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
Error during callback
java.lang.RuntimeException: java.io.IOException: Unix domain sockets are not supported on Windows
        at com.github.dockerjava.okhttp.UnixSocketFactory.createSocket(UnixSocketFactory.java:65)
        at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:257)
        at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:183)
        at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.java:224)
        at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.java:108)
        at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.java:88)
        at okhttp3.internal.connection.Transmitter.newExchange(Transmitter.java:169)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:41)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
        at okhttp3.RealCall.execute(RealCall.java:81)
        at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.execute(OkHttpInvocationBuilder.java:276)
        at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.lambda$executeAndStream$4(OkHttpInvocationBuilder.java:317)
Caused by: java.io.IOException: Unix domain sockets are not supported on Windows
        at com.github.dockerjava.okhttp.UnixDomainSocket.<init>(UnixDomainSocket.java:93)
        at com.github.dockerjava.okhttp.UnixSocketFactory$1.<init>(UnixSocketFactory.java:24)
        at com.github.dockerjava.okhttp.UnixSocketFactory.createSocket(UnixSocketFactory.java:24)
        ... 21 more
> Task :generateJooqClasses FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateJooqClasses'.
> java.io.IOException: Unix domain sockets are not supported on Windows

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
1 actionable task: 1 executed
```

I'll try to reproduce it in a different way

**Comment by strindberg** [comment]
Ah, sorry. I forgot about the extra configuration required on Windows. Basically, you need to have Docker running, and then specifying the the docker engine connection by setting a system variable (this is not necessary on Linux or MacOS):

    DOCKER_HOST="npipe:////./pipe/docker_engine"

**Comment by lukaseder** [comment]
Thanks for the hint. It's OK, I can reproduce the problem with the `CREATE TABLE` and the kotlin snippet in our own test environment.

**Comment by lukaseder** [comment]
OK, the way I see it, I guess since we're using reflection already to discover the presence of Jackson (currently), it will suffice to first attempt to call `com.fasterxml.jackson.module.kotlin.ExtensionsKt.jacksonObjectMapper` and load that particular `ObjectMapper`.

Users will still have to provide it in their own dependencies.

**Comment by lukaseder** [comment]
Our tests use H2, so the example still fails:

```
14:56:43,118 [ERROR] testConvertJSONIntoDataClassWithJackson(org.jooq.codegen.test.jackson.JacksonTest)  Time elapsed: 0.374 s  <<< ERROR!
org.jooq.exception.MappingException: An error ocurred when mapping record to class org.jooq.codegen.test.jackson.JacksonTest$Parent
        at org.jooq.codegen.test.jackson.JacksonTest.testConvertJSONIntoDataClassWithJackson(JacksonTest.kt:27)
Caused by: org.jooq.exception.DataTypeException: Error while mapping JSON to POJO using Jackson
        at org.jooq.codegen.test.jackson.JacksonTest.testConvertJSONIntoDataClassWithJackson(JacksonTest.kt:27)
Caused by: java.lang.reflect.InvocationTargetException
        at org.jooq.codegen.test.jackson.JacksonTest.testConvertJSONIntoDataClassWithJackson(JacksonTest.kt:27)
Caused by: com.fasterxml.jackson.module.kotlin.MissingKotlinParameterException:
Instantiation of [simple type, class org.jooq.codegen.test.jackson.JacksonTest$Child] value failed for JSON property name due to missing (therefore NULL) value for creator parameter name which is a non-nullable type
 at [Source: (String)"{"NAME":"c"}"; line: 1, column: 12] (through reference chain: org.jooq.codegen.test.jackson.JacksonTest$Child["name"])
        at org.jooq.codegen.test.jackson.JacksonTest.testConvertJSONIntoDataClassWithJackson(JacksonTest.kt:27)
```

This fixes it:

```kotlin
jsonObject(key("name").value(CHILD.NAME))
```

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.15.0. I won't backport this to 3.14, as these auto-configuration reflection magics always introduce new regression risk, breaking some other potential assumptions people may have.

Thanks again for your help, everyone.
